staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -23,57 +23,124 @@ import (
 	"io"
 	"net"
 	"net/http"
+	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/golang/glog"
 	"github.com/pborman/uuid"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apiserver/pkg/apis/audit"
 	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
 	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/filters"
 	authenticationapi "k8s.io/client-go/pkg/apis/authentication/v1"
 )
 
-var _ http.ResponseWriter = &auditResponseWriter{}
-
-type auditResponseWriter struct {
-	http.ResponseWriter
-	out io.Writer
-	id  string
+type AuditBackend interface {
+	// Log one event. After returning the passed event is reused. If the backend asynchronously process the
+	// event, it has to make a deep copy before. Per audit ID it might be that LogAuditEvent is called up to
+	// two times.
+	LogAuditEvent(ev *audit.Event) error
 }
 
-func (a *auditResponseWriter) WriteHeader(code int) {
-	line := fmt.Sprintf("%s AUDIT: id=%q response=\"%d\"\n", time.Now().Format(time.RFC3339Nano), a.id, code)
-	if _, err := fmt.Fprint(a.out, line); err != nil {
-		glog.Errorf("Unable to write audit log: %s, the error is: %v", line, err)
+func updateEventFromRequest(ev *audit.Event, req *http.Request, ctx request.Context) error {
+	attribs, err := GetAuthorizerAttributes(ctx)
+	if err != nil {
+		return err
 	}
 
-	a.ResponseWriter.WriteHeader(code)
-}
+	ev.Verb = attribs.GetVerb()
+	ev.RequestURI = req.URL.RequestURI()
+	ev.SourceIP = utilnet.GetClientIP(req).String()
```

Comment:
Leaving a TODO there for now. Can/Should be fixed in a follow-up PR which adds the fields to the event type.